stream/storage/impl/src/main/java/org/apache/bookkeeper/stream/storage/impl/service/RangeStoreServiceImpl.java

```diff
@@ -128,10 +140,31 @@
         this.rootRange = failRequestStorageContainer;
         this.mgStore = failRequestStorageContainer;
         this.storeFactory = storeFactory;
+        this.clientManager = clientManager;
         this.rrStoreFactory = rrStoreFactory;
         this.mrStoreFactory = mrStoreFactory;
         this.tableStoreFactory = tableStoreFactory;
         this.tableStoreCache = new TableStoreCache(storeFactory, tableStoreFactory);
+        this.streamProperties = new ConcurrentHashMap<Long, StreamProperties>();
+    }
+
+    private CompletableFuture<StreamProperties> getStreamProperties(long streamId) {
+        final StreamProperties p = streamProperties.get(streamId);
+        if (p != null) {
+            return FutureUtils.value(p);
+        }
+
+        return clientManager.getStreamProperties(streamId)
```

Comment:
Probably you can do this.
```
GetStreamRequest streamRequest = GetStreamRequest.newBuilder().setStreamId(streamId).build();
getStream(streamRequest)
```
Or you can expose `RootRangeStoreImpl:getStreamProps(long streamId)`
The one you are using is.
Making a GRPC call, which will eventually land here and execute same code.
Moreover, if you choose to use clientManager, you do not have to cache stream properties, because clientManager already does that, though incorrectly. I think ConcurrentHashmap does not protect you against race-condition unless you use putIfAbsent() API.

Reply:
RootRange reads from local storage. The getStreamProps service call fails in the base impl. I tried that first. :)

I'll delete the cache. CHM would protect against corrupting the internals bit potentially overwrite one reference with another equivalent one.  Since I didn't avoid concurrent RPC due to the unlikelihood once we get to the put it's hard to care.